crates/subspace-service/src/dsn/import_blocks/root_blocks.rs

```diff
@@ -0,0 +1,249 @@
+    }
+
+    /// Return last root block known to DSN. We ask several peers for the highest root block
+    /// known to them. Target root block should be known to at least ROOT_BLOCK_CONSENSUS_THRESHOLD
+    /// among peer set with minimum size of ROOT_BLOCK_CONSENSUS_MIN_SET peers.
+    async fn get_last_root_block(&self) -> Result<RootBlock, Box<dyn Error>> {
+        trace!("Getting last root block...");
+        let peer_blocks = Arc::new(RefCell::new(HashMap::<PeerId, Vec<RootBlock>>::new()));
+
+        // Get random peers. Some of them could be bootstrap nodes with no support for
+        // request-response protocol for records root.
+        let get_peers_result = self
+            .dsn_node
+            .get_closest_peers(PeerId::random().into())
+            .await;
+
+        let mut peer_block_tasks = Vec::new();
+        // Acquire root blocks from peers.
+        match get_peers_result {
+            Ok(mut get_peers_stream) => {
+                while let Some(peer_id) = get_peers_stream.next().await {
+                    let peer_blocks = peer_blocks.clone();
+                    let dsn_node = self.dsn_node.clone();
+                    let peer_block_fut = async move {
+                        trace!(%peer_id, "get_closest_peers returned an item");
+
+                        let request_result = dsn_node
+                            .send_generic_request(
+                                peer_id,
+                                RootBlockRequest::LastRootBlocks {
+                                    root_block_number: ROOT_BLOCK_NUMBER_PER_REQUEST,
+                                },
+                            )
+                            .await;
+
+                        match request_result {
+                            Ok(RootBlockResponse { root_blocks }) => {
+                                trace!(
+                                    %peer_id,
+                                    root_blocks_number=%root_blocks.len(),
+                                    "Last root block request succeeded."
+                                );
+
+                                peer_blocks.borrow_mut().insert(peer_id, root_blocks);
+                            }
+                            Err(error) => {
+                                debug!(%peer_id, ?error, "Last root block request failed.");
+                            }
+                        };
+                    };
+
+                    peer_block_tasks.push(peer_block_fut);
+                }
+            }
+            Err(err) => {
+                warn!(?err, "get_closest_peers returned an error");
+
+                return Err(err.into());
+            }
+        }
+
+        // Driving future task forward.
+        for task in peer_block_tasks {
+            task.await;
+        }
+
+        let peer_blocks = Arc::try_unwrap(peer_blocks)
+            .expect("We manually waited for each other usage to be dropped.")
+            .into_inner();
```

Comment:
I believe this does the same trick and also utilizes `FuturesUnordered` concurency.
```suggestion
                let peer_blocks = self
            .dsn_node
            .get_closest_peers(PeerId::random().into())
            .await? // TODO: print logs on error
            .map(|peer_id| {
                trace!(%peer_id, "get_closest_peers returned an item");
                self.dsn_node.send_generic_request(
                    peer_id,
                    RootBlockRequest::LastRootBlocks {
                        root_block_number: ROOT_BLOCK_NUMBER_PER_REQUEST,
                    },
                )
            })
            .collect::<futures::future::FuturesUnordered<_>>()
            .await
            .filter_map(|request_result| match request_result {
                 Ok(RootBlockResponse { root_blocks }) => {
                     trace!(
                         %peer_id,
                         root_blocks_number=%root_blocks.len(),
                         "Last root block request succeeded."
                     );

                     Some((peer_id, root_blocks))
                 }
                 Err(error) => {
                     debug!(%peer_id, ?error, "Last root block request failed.");
                     None
                 }
            })
            .collect::<HashMap<PeerId, Vec<RootBlock>>>();
```

Reply:
The code is not identical, peer_id is missing in the `filter_map`. But you're correct about the `FuturesUnordered`

Reply:
It seems `FuturesUnordered` doesn't work with async blocks. I changed it to the `join_all` future which in turn will automatically go to the `FuturesOrdered` if required.

Reply:
How about now? It should be identical, in terms of actual logic, but also it would utilize `FuturesUnordered`